pkgs/development/python-modules/tree-sitter0_21/default.nix

```diff
@@ -30,6 +30,10 @@ buildPythonPackage rec {
 
   pythonImportsCheck = [ "tree_sitter" ];
 
+  # Needed explicitly for Python >= 3.12 as tree-sitter provides
+  # calls to distutils functions to compile language files
+  dependencies = lib.optionals (pythonAtLeast "3.12") [ distutils ];
```

Comment:
Please don't propagate `distutils`. It should go to `build-system` instead.

Reply:
Without distutils being propagated to downstream libraries, tree-sitter 0.21 does not work on Python 3.12. It isn't part of this package's build system, but rather part of the library's internal calls that other libraries using this one depend on. Removing it from dependencies and putting it in build-system would be incorrect and render this package non-functional again.

Reply:
Weird. I wonder why it works with python<=3.11 even though setuptools was not propagated?
Well, this time, it's okay.

Reply:
distutils was still included in the base library up through 3.11. It has been deprecated since forever, but it wasn't removed from the stdlib until 3.12. Since this library uses distutils internally (up through version 0.21) with direct calls, it's needed to bring it forward with this version of tree-sitter.